### PR TITLE
MGMT-12627: Add Multiple VIPs support for state machine validators

### PR DESCRIPTION
**Pull request by nmagnezi**
For backward compatibility, the affected state machine validations
should support a single VIP (API and Ingress).

Relevant validators, with their new names:
```
api-vip-defined     --> api-vips-defined
api-vip-valid       --> api-vips-valid
ingress-vip-defined --> ingress-vips-defined
ingress-vip-valid   --> ingress-vips-valid
```
Closes: [MGMT-12627](https://issues.redhat.com//browse/MGMT-12627) 
Implements: https://github.com/openshift/assisted-service/pull/4245

## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by nmagnezi** [comment]
/hold
we should merge https://github.com/openshift/assisted-service/pull/4568 first.

File: internal/cluster/validator.go [hunk 1]
```diff
@@ -116,8 +119,9 @@ func (v *clusterValidator) isMachineCidrEqualsToCalculatedCidr(c *clusterPreproc
 	if swag.BoolValue(c.cluster.UserManagedNetworking) {
 		return ValidationSuccess, "The Cluster Machine CIDR is not required: User Managed Networking"
 	}
-	if c.cluster.APIVip == "" && c.cluster.IngressVip == "" {
-		return ValidationPending, "The Machine Network CIDR, API virtual IP, or Ingress virtual IP is undefined."
+	if c.cluster.APIVip == "" && (c.cluster.APIVips == nil || len(c.cluster.APIVips) == 0) &&
```
Comment (ori-amizur) [inline internal/cluster/validator.go:122]: Length of nil slice is 0, so no need to check nil equality.

Comment (nmagnezi) [reply internal/cluster/validator.go]: Done

File: internal/cluster/validations/validations.go [hunk 2]
```diff
@@ -312,73 +315,349 @@ func ValidateVipDHCPAllocationWithIPv6(vipDhcpAllocation bool, machineNetworkCID
...
+	if apiVip != "" && len(apiVips) == 0 {
+		return []*models.APIVip{{IP: models.IP(apiVip), ClusterID: clusterId}}, nil
+	}
+	// Both APIVip and APIVips were provided.
+	if apiVip != "" && len(apiVips) > 0 && apiVip != string(apiVips[0].IP) {
+		return nil, errors.New("apiVIP must be the same as the first element of apiVIPs")
+	}
+	// APIVips were provided, but APIVip was not.
+	if apiVip == "" && apiVips != nil && len(apiVips) > 0 {
+		return nil, errors.New("request must include apiVIP alongside apiVIPs")
```
Comment (danielerez) [inline internal/cluster/validations/validations.go:329]: Why the new client actually should send both singular and plural variant? Isn't it enough for a new client to send only the plural one? I.e. don't we want to deprecate the singular variant?

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -508,8 +518,10 @@ func (b *bareMetalInventory) RegisterClusterInternal(
 			Href:                         swag.String(url.String()),
 			Kind:                         swag.String(models.ClusterKindCluster),
 			APIVip:                       params.NewClusterParams.APIVip,
```
Comment (danielerez) [inline internal/bminventory/inventory.go:520]: worth adding some comment about deprecating the singular variants for later on? maybe in the params struct?

File: internal/cluster/validator.go [hunk 4]
```diff
@@ -135,42 +139,75 @@ func (v *clusterValidator) isMachineCidrEqualsToCalculatedCidr(c *clusterPreproc
...
-	err := network.VerifyVip(c.cluster.Hosts, network.GetMachineCidrById(c.cluster, 0), c.cluster.APIVip, ApiVipName,
+
+	err = network.VerifyVip(c.cluster.Hosts, network.GetMachineCidrById(c.cluster, 0), c.cluster.APIVip, ApiVipName,
 		true, v.log)
+	if err != nil {
+		multiErr = multierror.Append(multiErr, err)
+	}
+
+	for i := range c.cluster.APIVips {
+		if network.GetApiVipById(c.cluster, i) == c.cluster.APIVip {
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply internal/cluster/validator.go:198]: I used it because it already converts the from `model.IP` to string. The alternative is `string(c.cluster.APIVips[i].IP)` - should I change it?

File: internal/cluster/validations/validations.go (hunk 2 above)
Comment (nmagnezi) [reply internal/cluster/validations/validations.go:329]: It is best explained in @mkowalski's doc: https://github.com/openshift/assisted-service/pull/4245/files#diff-0f2f3122fee0aeba312f537e745da06d17e1c4588f7ea6177834bbf2bfa161ceR74

Comment (danielerez) [reply internal/cluster/validations/validations.go:329]: Yeah seen that, just didn't understand why getting a plural field isn't a good indication for a new client:/ I'll try to go over the doc again...

File: internal/cluster/validator.go [hunk 5]
```diff
@@ -135,42 +139,75 @@ func (v *clusterValidator) isMachineCidrEqualsToCalculatedCidr(c *clusterPreproc
...
+			// Don't check the same address twice.
+			continue
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply internal/cluster/validator.go]: Done. No longer addressing `c.cluster.APIVip` in state machine validators.

File: internal/cluster/validator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nmagnezi) [reply internal/cluster/validator.go]: Done, I'm no longer comparing to a single VIP here, but used string(c.cluster.APIVips[i].IP) directly in other places.

File: internal/cluster/validator.go [hunk 6]
```diff
@@ -135,42 +139,75 @@ func (v *clusterValidator) isMachineCidrEqualsToCalculatedCidr(c *clusterPreproc
...
+		if c.cluster.APIVip == "" {
+			return ValidationFailure, "API virtual IPs are defined, but API Vip is not."
 		}
+		return ValidationSuccess, "API virtual IPs are defined."
+	}
+
+	if swag.BoolValue(c.cluster.VipDhcpAllocation) {
 		if isDhcpLeaseAllocationTimedOut(c) {
-			return ValidationFailure, "The API virtual IP is undefined; IP allocation from the DHCP server timed out."
+			return ValidationFailure, "API virtual IPs are undefined; IP allocation from the DHCP server timed out."
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply internal/cluster/validator.go:167]: What do you mean? I didn't change the meaning, just changed it from singular to plural.

File: internal/cluster/validator.go [hunk 7]
```diff
@@ -135,42 +139,75 @@ func (v *clusterValidator) isMachineCidrEqualsToCalculatedCidr(c *clusterPreproc
...
-		if !validationStatusToBool(machineCidrDefined) {
-			return ValidationPending, "The Machine Network CIDR is undefined"
+	if swag.BoolValue(c.cluster.VipDhcpAllocation) && !validationStatusToBool(machineCidrDefined) {
+		return ValidationPending, "The Machine Network CIDR is undefined"
+	}
+
+	if c.cluster.APIVips != nil && len(c.cluster.APIVips) > 0 {
+		// Backwards compatibility
+		if c.cluster.APIVip == "" {
+			return ValidationFailure, "API virtual IPs are defined, but API Vip is not."
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply internal/cluster/validator.go]: As said in other responses, you are right. Removed.

File: internal/cluster/validator.go [hunk 8]
```diff
@@ -135,42 +139,75 @@ func (v *clusterValidator) isMachineCidrEqualsToCalculatedCidr(c *clusterPreproc
...
 	}
 	machineCidrDefined, _ := v.isMachineCidrDefined(c)
-	if swag.BoolValue(c.cluster.VipDhcpAllocation) {
-		if !validationStatusToBool(machineCidrDefined) {
-			return ValidationPending, "The Machine Network CIDR is undefined"
+	if swag.BoolValue(c.cluster.VipDhcpAllocation) && !validationStatusToBool(machineCidrDefined) {
+		return ValidationPending, "The Machine Network CIDR is undefined"
+	}
+
+	if c.cluster.APIVips != nil && len(c.cluster.APIVips) > 0 {
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply internal/cluster/validator.go]: Done.

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (nmagnezi) [reply internal/bminventory/inventory.go:520]: https://github.com/openshift/assisted-service/pull/4668

File: internal/bminventory/inventory.go [hunk 9]
```diff
@@ -2444,6 +2410,10 @@ func (b *bareMetalInventory) updateNetworkParams(params installer.V2UpdateCluste
 			updates["api_vip"] = ""
 			updates["ingress_vip"] = ""
 			cluster.MachineNetworks = []*models.MachineNetwork{}
+			emptyCluster := common.Cluster{Cluster: models.Cluster{ID: cluster.ID}}
+			if err = network.UpdateVipsTables(db, &emptyCluster, true, true); err != nil {
+				return err
+			}
```
Comment (mkowalski) [inline internal/bminventory/inventory.go:2416]: I know the original comment `// VIP DHCP mode has changed` is not yours, but could you please expand it to something like

```
// VIP DHCP mode has changed. Because of this we are removing all the previously configured VIPs. This is because we do not want the non-DHCP mode to transparently use VIPs previously obtained via DHCP, nor we want DHCP mode to be fed with manually provided VIPs to start with
```

It is clear that this conditional fires when VIP DHCP mode has changed, but unless you know internals, it's difficult to unfold why the cleanup is happening. I hope putting a comment here can make it clear to whoever is touching this code in the future

File: internal/cluster/cluster_test.go [hunk 10]
```diff
@@ -2081,10 +2092,11 @@ var _ = Describe("SetVipsData", func() {
 			}
 			err := capi.SetVipsData(ctx, &cluster, t.apiVip, t.ingressVip, t.clusterApiLease, t.clusterIngressLease, db)
 			Expect(err != nil).To(Equal(t.errorExpected))
-			var c common.Cluster
-			Expect(db.Take(&c, "id = ?", clusterId.String()).Error).ToNot(HaveOccurred())
+			c := getClusterFromDB(clusterId, db)
 			Expect(c.APIVip).To(Equal(t.expectedApiVip))
+			Expect(network.GetApiVipById(&c, 0)).To(Equal(t.expectedApiVip))
 			Expect(c.IngressVip).To(Equal(t.expectedIngressVip))
+			Expect(network.GetIngressVipById(&c, 0)).To(Equal(t.expectedIngressVip))
```
Comment (mkowalski) [inline internal/cluster/cluster_test.go:2099]: For the sake of completeness, could you please check here also for `len(c.APIVips)` and `len(c.IngressVips)`? This will make us sure that any deletion we perform on the DB is successful

Comment (nmagnezi) [reply internal/cluster/cluster_test.go:2099]: Sure

File: internal/bminventory/inventory.go (hunk 9 above)
Comment (nmagnezi) [reply internal/bminventory/inventory.go:2416]: How about this wording?
> We are removing all the previously configured VIPs because the VIP
> DHCP mode has changed. Doing so will prevent situations where:
> 1. non-DHCP mode use VIPs previously obtained via DHCP transparently.
> 2. DHCP mode to be fed with manually provided VIPs.

Comment (nmagnezi) [reply internal/bminventory/inventory.go:2418]: Done

File: internal/cluster/cluster_test.go (hunk 10 above)
Comment (nmagnezi) [reply internal/cluster/cluster_test.go]: Done

File: internal/cluster/validator.go [hunk 11]
```diff
@@ -213,48 +238,57 @@ func isVipDhcpAllocationAndOVN(cluster *common.Cluster) bool {
...
-
-	if c.cluster.IngressVip != "" {
-		return ValidationSuccess, "The Ingress virtual IP is defined."
-	}
-
+	machineCidrDefined, _ := v.isMachineCidrDefined(c)
 	if swag.BoolValue(c.cluster.VipDhcpAllocation) {
+		if !validationStatusToBool(machineCidrDefined) {
+			return ValidationPending, "The Machine Network CIDR is undefined"
+		}
```
Comment (mkowalski) [inline internal/cluster/validator.go:252]: Before this PR the order of validations is
1) check if machine cidr is defined
2) check if ingress vips are defined

After this PR, the order changes so
* we check if ingress VIPs are defined and return Success if yes
* only after this, we check if machine cidr is defined

So I can imagine a scenario in which you have Ingress VIP but you are missing Machine Networks. Previously this would fail the validation. With this change, the validation will pass. Am I right? If that's the case, I'm not sure it's desired behaviour.

On the other hand in the old `areApiVipsDefined` function the order was always reversed, so I'm not sure the original author ever thought about this...

Comment (nmagnezi) [reply internal/cluster/validator.go]: I changed it back to the original order

File: internal/cluster/statemachine.go [hunk 12]
```diff
@@ -61,8 +61,8 @@ func NewClusterStateMachine(th *transitionHandler) stateswitch.StateMachine {
 	})
 
 	var pendingConditions = stateswitch.And(If(IsMachineCidrDefined), If(isClusterCidrDefined), If(isServiceCidrDefined), If(IsDNSDomainDefined), If(IsPullSecretSet))
-	var vipsDefinedConditions = stateswitch.And(If(IsApiVipDefined), If(IsIngressVipDefined))
-	var requiredForInstall = stateswitch.And(If(IsMachineCidrEqualsToCalculatedCidr), If(IsApiVipValid), If(IsIngressVipValid), If(AllHostsAreReadyToInstall),
+	var vipsDefinedConditions = stateswitch.And(If(AreIngressVipsDefined), If(AreApiVipsDefined), If(AreIngressVipsDefined))
```
Comment (omertuc) [inline internal/cluster/statemachine.go:64]: Is `AreIngressVipsDefined` repeated on purpose?
